### PR TITLE
fix(rust): `oneOf` generation for client

### LINKED ISSUE #17869: [BUG] [RUST] Enum Varients Merging Incorrectly, breaking change
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

I'm receiving a new compilation error when using an email with two tuple variants that reduce to the same type. Unfortunately, I suspect the error is related to https://github.com/OpenAPITools/openapi-generator/pull/13970 .

I'm using the newtype pattern to validate the string with RAII. I have two types 
```rust
// My email type.
struct Email(String)
// My account name type
struct AccountName(String)
```

 Because I want my API route to access either an email or an account name using login, I have a third type:
```rust
#[derive(ToSchema, Debug, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[serde(untagged)]
enum AccountOrEmail {
    EmailAddress(Email),
    AccountName(AccountName),
}
```
(These enum varients may look identical but they're actually mutually exclusive and disambiguated during serde::deserialize via a `TryFrom` impl.) 

I'm using Utoipa to generate an `openapi.json` file for my Rust server, and I generate my client using the Docker image for openapi-generator. 

Using the Docker image tagged at `v7.2.0`, the `AccountOrEmail` type would generate as...

```rust
/*
 * {{Omitted from GitHub Issue}}
 *
 * {{Omitted from GitHub Issue}}
 *
 * The version of the OpenAPI document: 0.1.0
 * 
 * Generated by: https://openapi-generator.tech
 */

/// AccountOrEmail : When logging in, users can identify their account by either account name or email address, since both are unique to their account.



#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct AccountOrEmail {
}

impl AccountOrEmail {
    /// When logging in, users can identify their account by either account name or email address, since both are unique to their account.
    pub fn new() -> AccountOrEmail {
        AccountOrEmail {
        }
    }
}
```

This type is completely useless, but it _did_ compile. https://github.com/OpenAPITools/openapi-generator/pull/13970 sought to make the generated code actually useful (and I really benefit from the PR!). However, using the `latest` docker image, which I confirmed was published after the PR landed, the following code is generated:

```rust
/*
 * 
 * 
 * Generated by: https://openapi-generator.tech
 */

use super::String;



#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum AccountOrEmail {
    String(Box<String>),
}

impl Default for AccountOrEmail {
    fn default() -> Self {
        Self::String(Box::default())
    }
}
```

Note the spurious import of `use super::String`; no such file exists, and the name conspicuously conflicts with `std::string::String` in the default namespace. This code no longer compiles. Here's the console output:

```txt
   Compiling serde v1.0.196
   Compiling proc-macro2 v1.0.78
   Compiling quote v1.0.35
   Compiling syn v2.0.48
    Checking serde_json v1.0.113
    Checking serde_urlencoded v0.7.1
    Checking uuid v1.7.0
   Compiling serde_derive v1.0.196
    Checking reqwest v0.11.24
    Checking wack-http-client v1.0.0 (/Users/robbiemckinstry/workspace/{{omitted from GitHub issue}})
error[E0432]: unresolved import `super::String`
 --> crates/wack-http-client/src/models/account_or_email.rs:7:5
  |
7 | use super::String;
  |     ^^^^^^^^^^^^^ no `String` in `models`
  |
help: consider importing one of these items instead
  |
7 | use crate::models::AccountOrEmail::String;
  |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
7 | use serde::__private::de::Content::String;
  |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
7 | use serde_json::Value::String;
  |     ~~~~~~~~~~~~~~~~~~~~~~~~~
7 | use std::string::String;
  |     ~~~~~~~~~~~~~~~~~~~

For more information about this error, try `rustc --explain E0432`.
error: could not compile `wack-http-client` (lib) due to 1 previous error
```

##### openapi-generator version

I'm using the `latest` Docker image (accessed on 2/14/24). This appears to be a regression; using the `v7.2.0` image, the code will compile (but it won't be useful without [#13790](https://github.com/OpenAPITools/openapi-generator/pull/13970)).

##### OpenAPI declaration file content or url

 I paired down the generated specification to be the smaller possible valid spec that reproduces the issue:

```json
{
  "openapi": "3.0.3",
  "info": {
    "title": "Omitted",
    "description": "Omitted",
    "license": {
      "name": ""
    },
    "version": "0.1.0"
  },
  "servers": [
    {
      "url": "http://localhost:8080",
      "description": "Omitted"
    }
  ],
  "paths": {},
  "components": {
    "schemas": {
      "AccountName": {
        "type": "string",
        "description": "An account name."
      },
      "AccountOrEmail": {
        "oneOf": [
          {
            "$ref": "#/components/schemas/Email"
          },
          {
            "$ref": "#/components/schemas/AccountName"
          }
        ],
        "description": "Either an account name or an email."
      },
      "Email": {
        "type": "string",
        "description": "An email address."
      }
    },
    "responses": {}
  }
}
```

##### Generation Details

```bash
docker run --rm -v "$PWD:/local" \
  openapitools/openapi-generator-cli:latest generate \
  --skip-validate-spec \
  -i /local/target/debug/openapi.json \  # NOTE: Replace with the file provided above
  -g rust \
  --package-name "wack-http-client" \
  -o /local/crates/wack-http-client
```

##### Steps to reproduce

1. Copy the `openapi.json` file I included above into $PWD/openapi.json.
2. Run the following:

```bash
docker run --rm -v "$PWD:/local" \
  openapitools/openapi-generator-cli:latest generate \
  --skip-validate-spec \
  -i /local/openapi.json \  # NOTE: Replace with the file provided above
  -g rust \
  --package-name "foobar" \
  -o /local/foobar
```

##### Related issues/PRs

https://github.com/OpenAPITools/openapi-generator/pull/13970

I really hope this bug can be patched without reversing the above PR, because it's a much needed addition to the project 😅 

##### Suggest a fix

It looks like the two newtypes are getting incorrectly merged; instead, I believe the models for `AccountName` and `Email` should be generated and referenced in the enum.

### PR DESCRIPTION
**Pull request by DDtKey**
Closes #17869, #17896, #9497 and also includes unmerged #17898 (fix of `serde::rename` mappings)

Unfortunately it affects quite a lot of code, but we can see that only client-side models were affected by re-generation. 
I tried to split this PR to several, but they're really coupled and hard to create a chain of PRs (e.g #14898 & #17899 and issue with imports).

Other changes:
- imports are aligned between Rust codegens via`AbstractRustCodegen` (it used to be error-prone copy/paste with small differences for server & client)
- related improvement is described here: https://github.com/OpenAPITools/openapi-generator/pull/17915#issuecomment-1957881796
- fixed `number` representation: `f32` -> `f64`

### PR checklist
 
- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [X] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [X] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.1.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [X] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@frol @farcaller @richardwhiuk @paladinzh @jacob-pro

### CODE REVIEW

File: samples/client/others/rust/hyper/composed-oneof/src/models/obj_a.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+use crate::models;
```
Comment (DDtKey) [inline samples/client/others/rust/hyper/composed-oneof/src/models/obj_a.rs:1]: It's now aligned with server-side generator via `AbstractRustCodegen`. So a lot of sample files have only this change

**Comment by DDtKey** [comment]
`modules/openapi-generator/src/test/resources/3_0/oneOfArrayMapImport.yaml` covers issue described in #17869 pretty well

But JFYI, it generates the following enum against example in that PR:
```rust
/// Either an account name or an email.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum AccountOrEmail {
    /// An email address.
    Email(Box<String>),
    /// An account name.
    AccountName(Box<String>),
}
```

cc @RobbieMcKinstry

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java [hunk 2]
```diff
@@ -205,11 +206,45 @@ public RustClientCodegen() {
...
+    }
+
     @Override
     public ModelsMap postProcessModels(ModelsMap objs) {
         // Remove the discriminator field from the model, serde will take care of this
         for (ModelMap model : objs.getModels()) {
+            System.out.println("\nMODEL: \n\n");
+            System.out.println(model);
+            System.out.println("\n\n\n");
+
```
Comment (ctrlaltf24) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:245]:
```suggestion
```

**Review by ctrlaltf24 — COMMENTED** [review]
Thanks for the fixes @DDtKey ! Doesn't look like the samples compile as-is.
```bash
cd rust/client/others/rust
cargo check
```
nor does the cli compile
```
docker build .
```

Will review in-depth tomorrow to try to help get this merged ASAP. Thanks again!

**Comment by DDtKey** [comment]
There are missing changes, sorry

It definitely works with `cargo check` on the latest commit
Pushed

File: samples/client/others/rust/hyper/emptyObject/src/models/empty_object.rs [hunk 3]
```diff
@@ -8,6 +8,7 @@
  * Generated by: https://openapi-generator.tech
  */
 
+use crate::models;
```
Comment (DDtKey) [inline samples/client/others/rust/hyper/emptyObject/src/models/empty_object.rs:11]: A lot of client files contain only this change - it's now aligned with server-side generators (via `AbstractRustCodegen`).

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractRustCodegen.java [hunk 4]
```diff
@@ -232,6 +233,83 @@ public String sanitizeIdentifier(String name, CasingType casingType, String esca
         return name;
     }
 
+    @Override
+    public String getTypeDeclaration(Schema p) {
```
Comment (DDtKey) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractRustCodegen.java:237]: It was error-prone because almost copy-paste between rust-generators, so only extracted from server side and unified with client .

We can see that only `client` samples were affected by this PR.

**Comment by DDtKey** [comment]
PR is ready for review, cleaned-up the code a bit. 

- passing `cargo check`
- only `client` samples are affected by this PR (even taking unification with server into account)

**Comment by DDtKey** [comment]
PR is ready for review, cleaned-up the code a bit

**Comment by RobbieMcKinstry** [comment]
> `modules/openapi-generator/src/test/resources/3_0/oneOfArrayMapImport.yaml` covers issue described in #17869 pretty well
> 
> But JFYI, it generates the following enum against example in that PR:
> 
> ```rust
> /// Either an account name or an email.
> #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
> #[serde(untagged)]
> pub enum AccountOrEmail {
>     /// An email address.
>     Email(Box<String>),
>     /// An account name.
>     AccountName(Box<String>),
> }
> ```
> 
> cc @RobbieMcKinstry

That looks good to me! It's not obvious to me why the `String`s have to be boxed, but that's not even a minor concern for me. 👍🏻 

Thank you very much for this PR! ❤️

**Comment by DDtKey** [comment]
> It's not obvious to me why the Strings have to be boxed
Yeah, also find this confusing. But I don't want to change too much in this PR. I think it can be separate PR to clean-up Box usage

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java [hunk 5]
```diff
@@ -399,43 +450,9 @@ public String modelDocFileFolder() {
...
-            return schemaType;
-        }
-
-        if (languageSpecificPrimitives.contains(schemaType)) {
-            return schemaType;
-        }
-
-        // return fully-qualified model name
-        // crate::models::{{classnameFile}}::{{classname}}
-        return "crate::models::" + toModelName(schemaType);
```
Comment (ctrlaltf24) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:438]: This returned the fully-qualified model name before, yet don't see any changes in samples/ perhaps this code path was uncovered by the existing sample?

File: modules/openapi-generator/src/main/resources/rust/hyper/api.mustache
Comment (ctrlaltf24) [inline modules/openapi-generator/src/main/resources/rust/hyper/api.mustache:1]: Why these changes? Isn't it more robust to use fully qualified names?

File: modules/openapi-generator/src/main/resources/rust/reqwest/api.mustache
Comment (ctrlaltf24) [inline modules/openapi-generator/src/main/resources/rust/reqwest/api.mustache:1]: Why import from the model path rather than fully qualified?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java [hunk 6]
```diff
@@ -205,11 +204,63 @@ public RustClientCodegen() {
...
+                    .collect(Collectors.groupingBy(Map.Entry::getValue,
+                            Collectors.mapping(Map.Entry::getKey, Collectors.toList())
+                    ));
+
+            for (int i = 0; i < newOneOfs.size(); i++) {
+                CodegenProperty oneOf = newOneOfs.get(i);
+                Schema schema = schemas.get(i);
+
+                if (mappedNamesByRef.containsKey(schema.get$ref())) {
+                    String mappedName = mappedNamesByRef.get(schema.get$ref()).removeFirst();
```
Comment (ctrlaltf24) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:239]: I'm not very familiar with the codegen code, but why is there a removeFirst here?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java [hunk 7]
```diff
@@ -205,11 +204,63 @@ public RustClientCodegen() {
...
+                    oneOf.setBaseName(mappedName);
+                    oneOf.setName(toModelName(mappedName));
+                } else {
+                    String refName = ModelUtils.getSimpleRef(schema.get$ref());
+                    if (refName != null) {
```
Comment (ctrlaltf24) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:244]: Do we have a test that exercises this path?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java [hunk 8]
```diff
@@ -149,11 +146,13 @@ public RustClientCodegen() {
         typeMapping.clear();
         typeMapping.put("integer", "i32");
         typeMapping.put("long", "i64");
-        typeMapping.put("number", "f32");
+        typeMapping.put("number", "f64");
```
[the remark this answers is not in the document]
Comment (DDtKey) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:149]: It was wrong - number should be `f64`. 
And currently master branch already have breaking changes unfortunately - see #17869

This can be reverted though, but I thought it's nice to get fixed

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java (hunk 5 above)
Comment (DDtKey) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:438]: There are a lot of changes in sample related to imports, see my comment above:
https://github.com/OpenAPITools/openapi-generator/pull/17915#discussion_r1497538208

It's now aligned for Rust codegens

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ctrlaltf24) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:149]: Sounds good, we should just mention it somewhere.

File: modules/openapi-generator/src/main/resources/rust/hyper/api.mustache
Comment (DDtKey) [reply modules/openapi-generator/src/main/resources/rust/hyper/api.mustache:1]: It's aligned with server side codegen, I didn't want to affect server impl and sometimes it makes the code just a bit shorter.

File: modules/openapi-generator/src/main/resources/rust/reqwest/api.mustache
Comment (DDtKey) [reply modules/openapi-generator/src/main/resources/rust/reqwest/api.mustache:1]: The same answer, it's now aligned between rust codegens

File: modules/openapi-generator/src/main/resources/rust/hyper/api.mustache
Comment (DDtKey) [reply modules/openapi-generator/src/main/resources/rust/hyper/api.mustache:1]: It's compatible change, but unification of codegens made this less error-prone

File: modules/openapi-generator/src/main/resources/rust/reqwest/api.mustache
Comment (DDtKey) [reply modules/openapi-generator/src/main/resources/rust/reqwest/api.mustache:1]: It's compatible change, but unification of codegens made this less error-prone

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java (hunk 5 above)
Comment (DDtKey) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:438]: It's compatible change, but unification of codegens made this less error-prone

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java (hunk 7 above)
Comment (DDtKey) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:244]: It was added here: https://github.com/OpenAPITools/openapi-generator/pull/17915/commits/26bd3b2977377ab44091d7a256f704647c311b4b

Comment (DDtKey) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:244]: Sure, if we would remove that - quite a lot of sample code would get broken

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java (hunk 6 above)
Comment (DDtKey) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:239]: There is small comment above:

```
// Reverse mapped references to use as baseName for oneOF, but different keys may point to the same $ref.
// Thus, we group them by the value
```

Basically, we may have several references to the same type, but with different mapping names. So we just need to use all names for the reference

Test for such case was added in this commit: https://github.com/OpenAPITools/openapi-generator/commit/26bd3b2977377ab44091d7a256f704647c311b4b

**Review by ctrlaltf24 — APPROVED** [review]
concerns addressed, LGTM. Works with my openapi definition that motivated the original change.

Thanks @DDtKey !

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java (hunk 5 above)
Comment (DDtKey) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:438]: Particularly this codepath was mostly tested by `petstore`, for example in main, before this change: https://github.com/OpenAPITools/openapi-generator/blob/f1919042974ec17821d66a5529e4a8df5d3b3ac7/samples/client/petstore/rust/reqwest/petstore-async/src/models/pet.rs#L20

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java (hunk 6 above)
Comment (DDtKey) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:239]: I'll add a comment

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (DDtKey) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:149]: Added to the description of the PR. As well as mention of import changes

**Comment by wing328** [comment]
thanks for the PR. please review the build failure when you've time, e.g. https://github.com/OpenAPITools/openapi-generator/actions/runs/7996641152/job/21865620537?pr=17915

**Comment by DDtKey** [comment]
Ouch, sorry, my bad - there is no such method for the interface, but only for a particular type.

Fixed, build works as expected (at least locally)

**Comment by DDtKey** [comment]
Looks like `node0` in `ci/circleci` failed not because of the changes. Seems like flakiness?

**Comment by DDtKey** [comment]
Sorry, @wing328 any chance it will be merged soon? 

CI is green and I'd like to update #17931 once this one is merged 🙂

**Comment by wing328** [comment]
lgtm. let's give it a try

have a nice weekend
